lib/sycamore/sycamore/utils/markdown.py

```diff
@@ -0,0 +1,152 @@
+                continue
+            last = bbox
+        if type == "table":
+            s += render_table(cast(TableElement, elem))
+            continue
+        text = elem_text(elem).strip()
+        if not text:
+            continue
+        text = text.replace("\n", " ")
+        if type == "title":
+            s += f"\n# {text}\n\n"
+        elif type == "section-header":
+            s += f"\n## {text}\n\n"
+        elif type == "list-item":
+            tup = elem.data.get("_listctx")
+            if tup:
+                indent, bullet = tup
+                n = len(indent) + len(bullet)
+                t = text[n:]
+                if not t[0].isspace():
+                    t = " " + t
+                s += f"{indent}-{t}\n"
+            else:
+                s += f"- {text}\n"
+        elif type in ("caption", "footnote"):
+            s += f"\n{text}\n\n"
+        else:
+            s += text + "\n"
+    return s
+
+
+def render_table(elem: TableElement) -> str:
+    table = elem.table
+    if not table:
+        return ""
+    nrow = table.num_rows
+    ncol = table.num_cols
+    cells = table.cells
+    matrix = [[""] * ncol for _ in range(nrow)]
+    hdr_max = -1
+    for cell in cells:
+        hdr = cell.is_header
+        for row in cell.rows:
+            if hdr:
+                hdr_max = max(hdr_max, row)
+            for col in cell.cols:
+                if cell.content:
+                    matrix[row][col] = cell.content
+    sep = "| " + " | ".join(["-----" for _ in range(ncol)]) + " |\n"
+    s = "\n"
+    if hdr_max < 0:
+        s += "|  " * ncol + "|\n"
+        s += sep
+    for row in range(nrow):
+        s += "| " + " | ".join(matrix[row]) + " |\n"
+        if row == hdr_max:
+            s += sep
+    return s + "\n"
+
+
+def label_lists(elems: list[Element]) -> None:
+    """
+    Assumes elements are already sorted in reading order.
+    """
+    num = len(elems)
+    idx = 0
+    while idx < num:
+        if elem_type(elems[idx]) == "list-item":
+            end = idx + 1
+            while (end < num) and (elem_type(elems[idx]) == "list-item"):
+                end += 1
+            n = end - idx
+            if n > 1:
+                indent, bullet = common_prefix(elems, idx, end)
+                if indent or bullet:
+                    for i in range(idx, end):
+                        elems[i].data["_listctx"] = (indent, bullet)
+            idx = end
+        else:
+            idx += 1
+
+
+def common_prefix(elems: list[Element], beg: int, end: int) -> tuple[str, str]:
+    ary = [elem_text(elems[i]) for i in range(beg, end)]
+    num = len(ary)
+    lng = len(min(ary, key=len))
+    indent = ""
+    bullet = ""
+    for pos in range(0, lng):
+        ch = ary[0][pos]
+        same = True
+        for idx in range(1, num):
+            if ch != ary[idx][pos]:
+                same = False
+                break
+        if (not same) or ch.isalnum():
+            break
```

Comment:
What about lists like 
1. foo
2. bar

Reply:
We don't currently remove common prefixes from ordered lists.  Basically, these render with bullets in front of the numbers.  Similar things happen with hierarchical lists.  They actually don't look too bad.  Not a lot of great options for hierarchical lists.  We can add recognition/rendering for simple numerical lists, but I'd rather do this later.  I'd expect it to be a bit complicated and also to be very susceptible to OCR and partitioning errors.